streams/src/test/java/org/apache/kafka/streams/processor/internals/RecordCollectorTest.java

```diff
@@ -62,206 +64,203 @@
-        collector.send("topic1", "82", "0", headers, null, stringSerializer, stringSerializer, streamPartitioner);
-        collector.send("topic1", "244", "0", headers, null, stringSerializer, stringSerializer, streamPartitioner);
-
-        collector.send("topic1", "245", "0", null, null, stringSerializer, stringSerializer, streamPartitioner);
+        collector.send(topic, "3", "0", null, null, stringSerializer, stringSerializer, streamPartitioner);
+        collector.send(topic, "9", "0", null, null, stringSerializer, stringSerializer, streamPartitioner);
+        collector.send(topic, "27", "0", null, null, stringSerializer, stringSerializer, streamPartitioner);
+        collector.send(topic, "81", "0", null, null, stringSerializer, stringSerializer, streamPartitioner);
+        collector.send(topic, "243", "0", null, null, stringSerializer, stringSerializer, streamPartitioner);
+        collector.send(topic, "28", "0", headers, null, stringSerializer, stringSerializer, streamPartitioner);
+        collector.send(topic, "82", "0", headers, null, stringSerializer, stringSerializer, streamPartitioner);
+        collector.send(topic, "244", "0", headers, null, stringSerializer, stringSerializer, streamPartitioner);
+        collector.send(topic, "245", "0", null, null, stringSerializer, stringSerializer, streamPartitioner);
 
         final Map<TopicPartition, Long> offsets = collector.offsets();
 
-        assertEquals((Long) 4L, offsets.get(new TopicPartition("topic1", 0)));
-        assertEquals((Long) 2L, offsets.get(new TopicPartition("topic1", 1)));
-        assertEquals((Long) 0L, offsets.get(new TopicPartition("topic1", 2)));
-    }
+        assertEquals((Long) 4L, offsets.get(new TopicPartition(topic, 0)));
+        assertEquals((Long) 2L, offsets.get(new TopicPartition(topic, 1)));
+        assertEquals((Long) 0L, offsets.get(new TopicPartition(topic, 2)));
+        assertEquals(9, producer.history().size());
 
-    @Test
-    public void shouldNotAllowOffsetsToBeUpdatedExternally() {
-        final String topic = "topic1";
+        // should not all offsets to be modified
         final TopicPartition topicPartition = new TopicPartition(topic, 0);
+        assertThrows(UnsupportedOperationException.class, () -> offsets.put(topicPartition, 50L));
+    }
 
-        final RecordCollectorImpl collector = new RecordCollectorImpl(
+    @Test
+    public void shouldUpdateOffsetsUponCompletion() {
+        final RecordCollector collector = new RecordCollectorImpl(
             taskId,
             streamsConfig,
             logContext,
             streamsMetrics,
-            null,
-            id -> new MockProducer<>(cluster, true, new DefaultPartitioner(), byteArraySerializer, byteArraySerializer)
+            consumer,
+            id -> new MockProducer<>(cluster, false, new DefaultPartitioner(), byteArraySerializer, byteArraySerializer)
         );
 
+        final Map<TopicPartition, Long> offsets = collector.offsets();
```

Comment:
q: so the offsets map is ever-changing with the underlying map inside collector? And this is the correct behavior as we don't need to create a deep-copy every time?

Reply:
This is the old behavior. But I think it is more appropriate to not making a view but a deep-copy, will change.